fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/lowcardinality/DecodeCollector.java

```diff
@@ -0,0 +1,609 @@
+    private static final Logger LOG = LogManager.getLogger(DecodeCollector.class);
+
+    public static final Set<String> LOW_CARD_AGGREGATE_FUNCTIONS =
+            Sets.newHashSet(FunctionSet.COUNT, FunctionSet.MULTI_DISTINCT_COUNT, FunctionSet.MAX, FunctionSet.MIN,
+                    FunctionSet.APPROX_COUNT_DISTINCT);
+
+    public static final Set<String> LOW_CARD_FUNCTIONS =
+            ImmutableSet.of(FunctionSet.APPEND_TRAILING_CHAR_IF_ABSENT, FunctionSet.CONCAT, FunctionSet.CONCAT_WS,
+                    FunctionSet.HEX, FunctionSet.LEFT, FunctionSet.LIKE, FunctionSet.LOWER, FunctionSet.LPAD,
+                    FunctionSet.LTRIM, FunctionSet.REGEXP_EXTRACT, FunctionSet.REGEXP_REPLACE, FunctionSet.REPEAT,
+                    FunctionSet.REPLACE, FunctionSet.REVERSE, FunctionSet.RIGHT, FunctionSet.RPAD, FunctionSet.RTRIM,
+                    FunctionSet.SPLIT_PART, FunctionSet.SUBSTR, FunctionSet.SUBSTRING, FunctionSet.SUBSTRING_INDEX,
+                    FunctionSet.TRIM, FunctionSet.UPPER, FunctionSet.IF);
+
+
+    // These fields are the same as the fields in the DecodeContext,
+    // the difference: these fields store all string information, the
+    // DecodeContext only stores the ones that need to be optimized.
+    private final Map<Operator, DecodeNodeInfo> allOperatorDecodeInfo = Maps.newIdentityHashMap();
+
+    private final Map<Integer, ColumnDict> globalDicts = Maps.newHashMap();
+
+    private final Map<Integer, List<ScalarOperator>> stringExpressions = Maps.newHashMap();
+
+    private final Map<Integer, List<CallOperator>> stringAggregateExpressions = Maps.newHashMap();
+
+    private final Map<Integer, ScalarOperator> stringRefToDefineExprMap = Maps.newHashMap();
+
+    // string column use counter, 0 meanings decoded immediately after it was generated.
+    // for compute global dict define expressions
+    private final Map<Integer, Integer> expressionStringRefCounter = Maps.newHashMap();
+
+    private final List<Integer> scanStringColumns = Lists.newArrayList();
+
+    public void collect(OptExpression root, DecodeContext context) {
+        collectImpl(root, null);
+        initContext(context);
+    }
+
+    private void initContext(DecodeContext context) {
+        // choose the profitable string columns
+        for (Integer cid : scanStringColumns) {
+            if (expressionStringRefCounter.getOrDefault(cid, 0) > 1) {
+                context.allStringColumns.add(cid);
+                continue;
+            }
+            List<ScalarOperator> dictExprList = stringExpressions.getOrDefault(cid, Collections.emptyList());
+            long allExprNum = dictExprList.size();
+            long worthless = dictExprList.stream().filter(ScalarOperator::isColumnRef).count();
+            if (worthless == 0 && allExprNum != 0) {
+                context.allStringColumns.add(cid);
+            } else if (allExprNum > worthless && allExprNum >= worthless * 2) {
```

Comment:
please check this branch condition and add some comments.

Reply:
updated